File: testsuite/OfflineExamples/WireframeOfflineExample/src/WireframeOfflineExample.java
// Java classes
import java.io.File;

// VSDK classes
import vsdk.toolkit.common.linealAlgebra.Matrix4x4;     // Model elements
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.environment.Camera;
import vsdk.toolkit.environment.geometry.volume.Box;
import vsdk.toolkit.environment.scene.SimpleBody;
import vsdk.toolkit.environment.scene.SimpleScene;
import vsdk.toolkit.media.Calligraphic2DBuffer;         // I/O artifacts
import vsdk.toolkit.media.RGBImage;
import vsdk.toolkit.io.geometry.EnvironmentPersistence; // Persistence elements
import vsdk.toolkit.io.image.ImagePersistence;
import vsdk.toolkit.render.WireframeRenderer;           // Processing elements

/**
This example program is the most fundamental computer graphics example in
VitralSDK that does not depend on any external libraries to generate an image
from a 3D scene. Note that it is based on:
  - A wireframe model imported from an external .obj file (requires a
    running platform with support to file systems, networked or "inline in
    code" reader).
  - Simple camera model and calligraphic renderer (100% java / Vitral SDK
    implementation)
  - Raster output using Vitral SDK image and Bresenham's line algorithm
  - This particular program exports the resulting image in a file (requires
    a platform supporting I/O to files).
*/
public class WireframeOfflineExample {
    private Camera camera;
    private SimpleScene scene;

    /**
    Note the starting simplicity of this constructor. Compare it against
    the Swing+JOGL version of this same example (VSDKExamples/WireframeExample)
    and recall that the only difference is that the offline example must
    update the virtual viewport space to the camera (in the interactive
    version this is done on a Swing/JOGL controlled callback function).
    */
    public WireframeOfflineExample() {
        createModel();
        camera.updateViewportResize(640, 480);
    }

    private void createModel()
    {
        //-----------------------------------------------------------------
        camera = new Camera();
        Matrix4x4 R = new Matrix4x4();

        camera.setPosition(new Vector3D(7, -4, 4));
        R.eulerAnglesRotation(Math.toRadians(140), Math.toRadians(-30), 0);
        camera.setNearPlaneDistance(0.001);
        camera.setFarPlaneDistance(100);
        camera.setRotation(R);

        //-----------------------------------------------------------------
        String sceneFile = "../../../etc/geometry/cow.obj";
        scene = new SimpleScene();

        try {
            EnvironmentPersistence.importEnvironment(new File(sceneFile), scene);
        }
        catch ( Exception ex ) {
            System.err.println("Failed to read file");
            ex.printStackTrace();
        }

        //-----------------------------------------------------------------
        SimpleBody b;
        Box box;

        b = new SimpleBody();
        box = new Box(1, 1, 1);
        b.setGeometry(box);
        b.setPosition(new Vector3D(1, 2, 3));
        scene.addBody(b);
    }

    public void rasterOutput(Calligraphic2DBuffer lineSet) {
        RGBImage outputImageRasterViewport;

        //- (1/2) line rasterization in to output image -------------------
        double xt = camera.getViewportXSize();
        double yt = camera.getViewportYSize();

        outputImageRasterViewport = new RGBImage();
        outputImageRasterViewport.init((int)xt, (int)yt);

        lineSet.exportRgbImage(outputImageRasterViewport);

        lineSet.init(); // leaves buffer ready for next frame

        //- (2/2) Image result transfer to output file --------------------
        ImagePersistence.exportJPG(
            new File("output.jpg"), outputImageRasterViewport);
    }

    public static void main (String[] args) {
        WireframeOfflineExample instance = new WireframeOfflineExample();
        Calligraphic2DBuffer lineSet;

        lineSet = new Calligraphic2DBuffer();
        WireframeRenderer.execute(
            lineSet, instance.scene.getSimpleBodies(), instance.camera);
        instance.rasterOutput(lineSet);
    }

}

File: cpp/testsuite/OfflineExamples/WireframeOfflineExample/src/WireframeOfflineExample.cpp
#include <cmath>
#include <cstdio>

#include "java/io/File.h"
#include "java/lang/String.h"
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/common/linealAlgebra/Matrix4x4d.h"
#include "vsdk/toolkit/common/linealAlgebra/Vector3Dd.h"
#include "vsdk/toolkit/media/Calligraphic2DBuffer.h"
#include "vsdk/toolkit/media/RGBImageUncompressed.h"
#include "vsdk/toolkit/media/RGBPixel.h"
#include "vsdk/toolkit/environment/geometry/volume/Arrow.h"
#include "vsdk/toolkit/environment/camera/Camera.h"
#include "vsdk/toolkit/environment/scene/SimpleBody.h"
#include "vsdk/toolkit/environment/scene/SimpleScene.h"
#include "vsdk/toolkit/io/image/ImagePersistence.h"
#include "vsdk/toolkit/io/geometry/EnvironmentPersistence.h"
#include "vsdk/toolkit/render/hiddenLine/HiddenLineRenderer.h"
#include "vsdk/toolkit/render/hiddenLine/WireframeRenderer.h"
#include "vsdk/toolkit/render/raster/Rasterizer2D.h"
static void
appendLineSet(Calligraphic2DBuffer* destination, const Calligraphic2DBuffer* source)
{
    if (destination == 0 || source == 0) {
        return;
    }

    for (int j = 0; j < source->getNumLines(); j++) {
        Vector3Dd* e0 = source->get2DLinePoint0(j);
        Vector3Dd* e1 = source->get2DLinePoint1(j);
        if (e0 != 0 && e1 != 0) {
            destination->add2DLine(*e0, *e1);
        }
        delete e0;
        delete e1;
    }
}

static void
rasterOutput(Camera* camera, Calligraphic2DBuffer* lineSet, const java::String& outputFile)
{
    RGBImageUncompressed outputImage;
    outputImage.init((int)camera->getViewportXSize(), (int)camera->getViewportYSize());

    RGBPixel white((char)255, (char)255, (char)255);
    const double xt = outputImage.getXSize();
    const double yt = outputImage.getYSize();

    for (int j = 0; j < lineSet->getNumLines(); j++) {
        Vector3Dd* e0 = lineSet->get2DLinePoint0(j);
        Vector3Dd* e1 = lineSet->get2DLinePoint1(j);
        if (e0 == 0 || e1 == 0) {
            delete e0;
            delete e1;
            continue;
        }

        int x0 = (int)((xt - 1) * ((e0->x() + 1.0) / 2.0));
        int y0 = (int)((yt - 1) * (1.0 - ((e0->y() + 1.0) / 2.0)));
        int x1 = (int)((xt - 1) * ((e1->x() + 1.0) / 2.0));
        int y1 = (int)((yt - 1) * (1.0 - ((e1->y() + 1.0) / 2.0)));
        Rasterizer2D::drawLine(&outputImage, x0, y0, x1, y1, white);

        delete e0;
        delete e1;
    }

    ImagePersistence::exportPNG(java::File(outputFile.c_str()), &outputImage);
}

int
main(int argc, char** argv)
{
    java::String sceneFile = "../../../../etc/geometry/cow.obj";
    java::String outputFile = "output.png";
    if (argc > 1 && argv[1] != 0 && java::String(argv[1]).size() > 0) {
        outputFile = argv[1];
    }

    Camera camera;
    Matrix4x4d R;
    const double pi = std::acos(-1.0);
    camera.setPosition(Vector3Dd(7, -4, 4));
    R = R.eulerAnglesRotation(140.0 * pi / 180.0, -30.0 * pi / 180.0, 0);
    camera.setNearPlaneDistance(0.001);
    camera.setFarPlaneDistance(100);
    camera.setRotation(R);
    camera.updateViewportResize(1024, 768);
    SimpleScene scene;
    EnvironmentPersistence::importEnvironment(java::File(sceneFile.c_str()), &scene);
    if (scene.getSimpleBodies().size() == 0) {
        std::fprintf(stderr, "Failed to load scene file: %s\n", sceneFile.c_str());
        return 1;
    }

    SimpleBody* arrowBody = new SimpleBody();
    arrowBody->setGeometry(new Arrow(1.0, 0.5, 0.15, 0.3));
    arrowBody->setRotation(Matrix4x4d().axisRotation(-30.0 * pi / 180.0, 1.0, 0.0, 0.0));
    arrowBody->setPosition(Vector3Dd(1, 2, 1.2));
    arrowBody->setScale(Vector3Dd(2, 2, 2));
    scene.addBody(arrowBody);

    java::ArrayList<SimpleBody*> wireframeBodies;
    java::ArrayList<SimpleBody*>& sceneBodies = scene.getSimpleBodies();
    for (long int i = 0; i < sceneBodies.size(); i++) {
        SimpleBody* body = sceneBodies.get(i);
        if (body != 0 && body != arrowBody) {
            wireframeBodies.add(body);
        }
    }

    java::ArrayList<SimpleBody*> hiddenLineBodies;
    hiddenLineBodies.add(arrowBody);

    Calligraphic2DBuffer cowWireframe;
    Calligraphic2DBuffer arrowVisibleContour;
    Calligraphic2DBuffer arrowVisibleNonContour;
    Calligraphic2DBuffer arrowHidden;
    Calligraphic2DBuffer finalLineSet;

    WireframeRenderer::execute(&cowWireframe, wireframeBodies, &camera);
    HiddenLineRenderer::executeAppelAlgorithm(
        hiddenLineBodies,
        &camera,
        &arrowVisibleContour,
        &arrowVisibleNonContour,
        &arrowHidden);

    appendLineSet(&finalLineSet, &cowWireframe);
    appendLineSet(&finalLineSet, &arrowVisibleContour);
    appendLineSet(&finalLineSet, &arrowVisibleNonContour);
    rasterOutput(&camera, &finalLineSet, outputFile);

    std::printf("Resulting image has been written to \"%s\"\n", outputFile.c_str());
    return 0;
}
